File: javascript/ql/src/experimental/Security/CWE-918/examples/SsrfIpv6TransitionIncompleteGuardBad.js
const isPrivate = require('private-ip');
const fetch = require('node-fetch');

// BAD: `private-ip` classifies the textual IPv4 form only, so it returns false
// for `::ffff:169.254.169.254`. The guard treats the wrapped internal address as
// public, but the request still reaches the metadata endpoint.
async function validateUrlHost(host) {
  if (isPrivate(host)) {
    throw new Error('blocked private host');
  }
  return fetch('http://' + host + '/');
}

module.exports = { validateUrlHost };

File: csharp/ql/src/experimental/CWE-918/examples/SsrfIpv6TransitionIncompleteGuardBad.cs
using System;
using System.Net;
using System.Net.Http;
using System.Threading.Tasks;

public class BadFetcher
{
    // BAD: a hand-written RFC 1918 / loopback / metadata denylist matched against the
    // textual host. The embedded IPv4 inside `::ffff:169.254.169.254` is never seen, so a
    // transition-wrapped internal address is classified as public and the request reaches it.
    private static bool IsPrivateHost(string host)
    {
        return host == "127.0.0.1"
            || host == "169.254.169.254"
            || host.StartsWith("10.")
            || host.StartsWith("192.168")
            || host.StartsWith("172.16");
    }

    public static async Task<string> FetchAsync(string host)
    {
        if (IsPrivateHost(host))
        {
            throw new Exception("blocked internal host");
        }

        using var client = new HttpClient();
        return await client.GetStringAsync("http://" + host + "/");
    }
}
